Wait for APB controller's clean external gateway ECMP routes before resuming default net controller startup

jordigilh opened the pull request:
Add logic to wait for APB controller's clean external gateway ECMP routes to be finished in the default network controller before resuming the execution of the `Run()` method.

Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.

@jcaamano PTAL.

jcaamano commented:
> Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.

Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.

In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.

jcaamano commented on go-controller/pkg/ovn/default_network_controller.go:
```diff
@@ -402,6 +402,15 @@ func (oc *DefaultNetworkController) Run(ctx context.Context) error {
+	oc.wg.Add(1)
+	go func() {
+		defer oc.wg.Done()
+		oc.apbExternalRouteController.Run(1)
+	}()
+
+	// Wait to finish repairing the external gateway ECMP routes before resuming
+	oc.apbExternalRouteController.RepairWaitGroup.Wait()
```
Any chance the repair could be run by itself outside of `apbExternalRouteController.Run`? I guess there is no issue to do it like its done here but it would be more straight forward an cleaner.

jordigilh commented:
> > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> 
> Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> 
> In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.



> > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> 
> Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> 
> In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.

The order matters in this case because the egressgw logic is embedded in the namespace and pod controllers, or the logic that watches for that in the [default network controller](https://github.com/ovn-org/ovn-kubernetes/blob/adcf68be2cfaf13ff88bf9479c90794f95318063/go-controller/pkg/ovn/pods.go#L236), so there is a potential race condition where the repair is cleaning up a pod that is not listed while the pod is added back while processing the loop that deletes its references. Basically this sequence:
1) Routine 1 processes the repair function and fetches all pods
2) while routine 1 is looping through the pods, one of the deleted pods is added again and routine 2 starts processing it as part of the pod controller in ovn
3) Routine 1 is preempted and freezes execution.
3) Routine 2 manages to complete processing and deletes the pod from the event queue.
4) Routine 1 resumes execution and deletes reference to the pod

Here the issue is that both APB and egressgw share the same cache and functionality. The moment the egressgw is removed in favour of APB, this race condition will not happen.

jcaamano commented:
> > > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> > 
> > 
> > Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> > In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.
> 
> > > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> > 
> > 
> > Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> > In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.
> 
> The order matters in this case because the egressgw logic is embedded in the namespace and pod controllers, or the logic that watches for that in the [default network controller](https://github.com/ovn-org/ovn-kubernetes/blob/adcf68be2cfaf13ff88bf9479c90794f95318063/go-controller/pkg/ovn/pods.go#L236), so there is a potential race condition where the repair is cleaning up a pod that is not listed while the pod is added back while processing the loop that deletes its references. Basically this sequence:
> 
> 1. Routine 1 processes the repair function and fetches all pods
> 2. while routine 1 is looping through the pods, one of the deleted pods is added again and routine 2 starts processing it as part of the pod controller in ovn
> 3. Routine 1 is preempted and freezes execution.
> 4. Routine 2 manages to complete processing and deletes the pod from the event queue.
> 5. Routine 1 resumes execution and deletes reference to the pod
> 
> Here the issue is that both APB and egressgw share the same cache and functionality. The moment the egressgw is removed in favour of APB, this race condition will not happen.

What this tells me though is that the repair logic needs to run before any watchers are started and any controllers are run. And how we usually organize this for clarity, is that things that need to run before the watchers are started go into the `func (oc *DefaultNetworkController) Init() ` method, and not in the `func (oc *DefaultNetworkController) Run(ctx context.Context)` method.

That's what I asked above if there was any chance the repair could be independent of the controller Run method so that we can place it in that Init method. Otherwise, if someone else does a repair logic for another controller in the future in a similar way as part of its Run method and it conflicts with this other repair logic, then we have contention for which runs first.

jordigilh commented:
> > > > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> > > 
> > > 
> > > Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> > > In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.
> > 
> > 
> > > > Also moved the initialization of the APB controller to be the first thing to do to ensure the north bound information is updated before all other controllers kickoff.
> > > 
> > > 
> > > Is the a functional difference on this order? Ideally there should not be any. If there is, it kinda smells like a design issue.
> > > In any case, could you add a comment detailing specifically why the controller must be started first? Otherwise, someone might just move it around.
> > 
> > 
> > The order matters in this case because the egressgw logic is embedded in the namespace and pod controllers, or the logic that watches for that in the [default network controller](https://github.com/ovn-org/ovn-kubernetes/blob/adcf68be2cfaf13ff88bf9479c90794f95318063/go-controller/pkg/ovn/pods.go#L236), so there is a potential race condition where the repair is cleaning up a pod that is not listed while the pod is added back while processing the loop that deletes its references. Basically this sequence:
> > 
> > 1. Routine 1 processes the repair function and fetches all pods
> > 2. while routine 1 is looping through the pods, one of the deleted pods is added again and routine 2 starts processing it as part of the pod controller in ovn
> > 3. Routine 1 is preempted and freezes execution.
> > 4. Routine 2 manages to complete processing and deletes the pod from the event queue.
> > 5. Routine 1 resumes execution and deletes reference to the pod
> > 
> > Here the issue is that both APB and egressgw share the same cache and functionality. The moment the egressgw is removed in favour of APB, this race condition will not happen.
> 
> What this tells me though is that the repair logic needs to run before any watchers are started and any controllers are run. And how we usually organize this for clarity, is that things that need to run before the watchers are started go into the `func (oc *DefaultNetworkController) Init() ` method, and not in the `func (oc *DefaultNetworkController) Run(ctx context.Context)` method.
> 
> That's what I asked above if there was any chance the repair could be independent of the controller Run method so that we can place it in that Init method. Otherwise, if someone else does a repair logic for another controller in the future in a similar way as part of its Run method and it conflicts with this other repair logic, then we have contention for which runs first.

I'll move it to the `init()` method and remove the `repair()` from the apb controller flow.

trozet commented:
Sorry for the delay. After reviewing this is a real problem. APB controller runs almost last and then tries to sync. It's sync will also try to remove things that shouldn't exist for the legacy path. The legacy path will already have been started and running by WatchPods and WatchNamespaces in the default controller. We do need this fix. @jordigilh can you please update this PR?

jordigilh commented:
> Sorry for the delay. After reviewing this is a real problem. APB controller runs almost last and then tries to sync. It's sync will also try to remove things that shouldn't exist for the legacy path. The legacy path will already have been started and running by WatchPods and WatchNamespaces in the default controller. We do need this fix. @jordigilh can you please update this PR?

Will do.

jordigilh commented:
@jcaamano @trozet please review. I have updated the PR to run the `repair()` function as part of the init() of the `DefaultNetworkController`.
